### ISSUE TITLE
[BUG] permission conflict between master and syndic for /var/cache/salt/master/proc (when using scheduling on master)

### ISSUE THREAD

**Issue by hurzhurz** [saltstack/salt#65095]
**Description**
I was trying to setup a scheduled job on a master that is also a syndic.
But running the job always fails and throws exceptions in the master log:
```
2023-08-31 07:39:49,450 [salt.utils.schedule:913 ][ERROR   ][49856] Unhandled exception running test.stdout_print
Traceback (most recent call last):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/schedule.py", line 827, in handle_func
    with salt.utils.files.fopen(proc_fn, "w+b") as fp_:
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/files.py", line 393, in fopen
    f_handle = open(*args, **kwargs)  # pylint: disable=resource-leakage
PermissionError: [Errno 13] Permission denied: '/var/cache/salt/master/proc/20230831073949422759'
2023-08-31 07:39:49,471 [salt.utils.process:998 ][ERROR   ][49856] An un-handled exception from the multiprocessing process 'Schedule(name=job1, jid=20230831073949422759)' was caught:
Traceback (most recent call last):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/process.py", line 993, in wrapped_run_func
    return run_func()
  File "/opt/saltstack/salt/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/schedule.py", line 960, in handle_func
    namespaced_event.destroy()
UnboundLocalError: local variable 'namespaced_event' referenced before assignment
```

The issue is that the master, that is running as user "salt", has no write permissions in /var/cache/salt/master/proc.
```
# ls -lah /var/cache/salt/master/proc
total 8.0K
drwxr-xr-x 2 root salt 4.0K Aug 31 11:09 .
drwxr-xr-x 9 salt salt 4.0K Aug 31 08:55 ..
```

While the master corrects the ownership on startup and sets the owner to "salt".
But then the syndic changes it to "root" on its startup (or on connect? not sure).
And as the dir permissions are set to 755 by default, the master has no write access anymore.

Changing the permissions of /var/cache/salt/master/proc to 775 seems to work as a workaround.
But I'm not sure if that completely solves it.

**Setup**
Minimal test setup: Two LXC containers with plain Ubuntu Jammy
First with salt-master (as master-of-masters), second with salt-master & salt-syndic. Installed using the Ubuntu packages from the salt repository.
Config is default, just added the syndic_master & order_masters settings as described here to get it running: https://docs.saltproject.io/en/latest/topics/topology/syndic.html

- [X] VM
- [X] container (LXC)
- [X] onedir packaging

**Steps to Reproduce the behavior**

Just add an scheduled job in the master config of a master&syndic host like that:
```
# cat /etc/salt/master.d/schedule.conf
schedule:
  job1:
    function: test.stdout_print
    seconds: 30
```

<details><summary>Debug Log</summary>

```
2023-08-31 07:39:47,274 [salt.master      :481 ][DEBUG   ][49103] Performing fileserver updates for items with an update interval of 60
2023-08-31 07:39:47,278 [salt.master      :463 ][DEBUG   ][49103] Updating roots fileserver cache
2023-08-31 07:39:47,279 [salt.master      :486 ][DEBUG   ][49103] Completed fileserver updates for items with an update interval of 60, waiting 60 seconds
2023-08-31 07:39:49,249 [salt.config      :2013][DEBUG   ][49089] Reading configuration from /etc/salt/master
2023-08-31 07:39:49,290 [salt.config      :2176][DEBUG   ][49089] Including configuration from '/etc/salt/minion.d/_schedule.conf'
2023-08-31 07:39:49,292 [salt.config      :2013][DEBUG   ][49089] Reading configuration from /etc/salt/minion.d/_schedule.conf
2023-08-31 07:39:49,294 [salt.config      :3616][DEBUG   ][49089] Using cached minion ID from /etc/salt/minion_id: test
2023-08-31 07:39:49,298 [salt.loader      :1073][DEBUG   ][49089] Grains refresh requested. Refreshing grains.
2023-08-31 07:39:49,299 [salt.config      :2013][DEBUG   ][49089] Reading configuration from /etc/salt/master
2023-08-31 07:39:49,322 [salt.config      :2176][DEBUG   ][49089] Including configuration from '/etc/salt/master.d/schedule.conf'
2023-08-31 07:39:49,323 [salt.config      :2013][DEBUG   ][49089] Reading configuration from /etc/salt/master.d/schedule.conf
2023-08-31 07:39:49,336 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'core' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,340 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'disks' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,342 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'extra' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,344 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'lvm' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,345 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'mdadm' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,348 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'minion_process' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,352 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'opts' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,354 [salt.loader.lazy :863 ][DEBUG   ][49089] Override  __utils__: <module 'salt.loaded.int.grains.zfs' from '/opt/saltstack/salt/lib/python3.10/site-packages/salt/grains/zfs.py'>
2023-08-31 07:39:49,355 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'zfs' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,392 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'zfs' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,394 [salt.utils.lazy  :99  ][DEBUG   ][49089] LazyLoaded zfs.is_supported
2023-08-31 07:39:49,406 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'local_cache' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,408 [salt.utils.lazy  :99  ][DEBUG   ][49089] LazyLoaded local_cache.clean_old_jobs
2023-08-31 07:39:49,416 [salt.loader.lazy :964 ][DEBUG   ][49089] The functions from module 'localfs' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,419 [salt.utils.lazy  :99  ][DEBUG   ][49089] LazyLoaded localfs.list_tokens
2023-08-31 07:39:49,420 [salt.utils.schedule:1762][DEBUG   ][49089] schedule: Job job1 was scheduled with jid_include, adding to cache (jid_include defaults to True)
2023-08-31 07:39:49,421 [salt.utils.schedule:1768][DEBUG   ][49089] schedule: Job job1 was scheduled with a max number of 1
2023-08-31 07:39:49,422 [salt.utils.schedule:1792][INFO    ][49089] Running scheduled job: job1 with jid 20230831073949422759
2023-08-31 07:39:49,425 [salt.utils.process:1145][DEBUG   ][49089] Subprocess Schedule(name=job1, jid=20230831073949422759) added
2023-08-31 07:39:49,427 [salt.utils.process:1156][DEBUG   ][49089] Subprocess Schedule(name=job1, jid=20230831073919227478) cleaned up
2023-08-31 07:39:49,428 [salt.utils.verify:426 ][DEBUG   ][49089] This salt-master instance has accepted 1 minion keys.
2023-08-31 07:39:49,443 [salt.loader.lazy :964 ][DEBUG   ][49856] The functions from module 'test' are being loaded by dir() on the loaded module
2023-08-31 07:39:49,445 [salt.utils.lazy  :99  ][DEBUG   ][49856] LazyLoaded test.stdout_print
2023-08-31 07:39:49,447 [salt.utils.schedule:821 ][DEBUG   ][49856] schedule.handle_func: adding this job to the jobcache with data {'id': 'test_master', 'fun': 'test.stdout_print', 'fun_args': [], 'schedule': 'job1', 'jid': '20230831073949422759', 'pid': 49856}
2023-08-31 07:39:49,450 [salt.utils.schedule:913 ][ERROR   ][49856] Unhandled exception running test.stdout_print
Traceback (most recent call last):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/schedule.py", line 827, in handle_func
    with salt.utils.files.fopen(proc_fn, "w+b") as fp_:
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/files.py", line 393, in fopen
    f_handle = open(*args, **kwargs)  # pylint: disable=resource-leakage
PermissionError: [Errno 13] Permission denied: '/var/cache/salt/master/proc/20230831073949422759'
2023-08-31 07:39:49,457 [salt.utils.event :315 ][DEBUG   ][49856] MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
2023-08-31 07:39:49,458 [salt.utils.event :316 ][DEBUG   ][49856] MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
2023-08-31 07:39:49,462 [salt.utils.event :823 ][DEBUG   ][49856] Sending event: tag = __schedule_return; data = {'cmd': '_return', 'id': 'test_master', 'fun': 'test.stdout_print', 'fun_args': [], 'schedule': 'job1', 'jid': 'req', 'pid': 49856, 'return': 'Unhandled exception running test.stdout_print', 'success': False, 'retcode': 254, '_stamp': '2023-08-31T07:39:49.462163'}
2023-08-31 07:39:49,466 [salt.transport.ipc:372 ][DEBUG   ][49856] Closing IPCMessageSubscriber instance
2023-08-31 07:39:49,469 [salt.transport.ipc:372 ][DEBUG   ][49856] Closing IPCMessageClient instance
2023-08-31 07:39:49,471 [salt.utils.process:998 ][ERROR   ][49856] An un-handled exception from the multiprocessing process 'Schedule(name=job1, jid=20230831073949422759)' was caught:
Traceback (most recent call last):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/process.py", line 993, in wrapped_run_func
    return run_func()
  File "/opt/saltstack/salt/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/schedule.py", line 960, in handle_func
    namespaced_event.destroy()
UnboundLocalError: local variable 'namespaced_event' referenced before assignment
```
</details>

**Versions Report**
<details><summary>salt --versions-report</summary>

```yaml
Salt Version:
          Salt: 3006.2

Python Version:
        Python: 3.10.12 (main, Aug  3 2023, 21:47:10) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.2
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.13.3
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: ubuntu 22.04.3 jammy
        locale: utf-8
       machine: x86_64
       release: 5.4.0-148-generic
        system: Linux
       version: Ubuntu 22.04.3 jammy
```
</details>

**Comment by dwoz** [comment]
@barneysowood thoughts?

**Comment by barneysowood** [comment]
@dwoz - I suspect the issue is that `salt-syndic` is running as root.. I've not used syndic in anger, but looking at the [docs](https://docs.saltproject.io/en/latest/topics/topology/syndic.html) the have a warning "The syndic process must be run as the same user as the syndic master"

Testing on 3006.x even if the master config has `user: salt` the `salt-syndic` runs as root. It's picking up `user` from the minion config. This is going to somewhat problematic as the syndic process is basically a minion, but should be running as the same user as the master. 

There are two further issues:

1. If you run the `salt-syndic` process as the "salt"` user, it can't access `/etc/salt/pki/minion` and various other minion related directories - to make it work those would have to have ownership changed/fixed.
2. Having a `salt-syndic` running as "salt" and a normal minion running as root on the same host (with the current config setup) won't work..

As an interim solution we may just need to document to options for people who want to run a syndic:

1. Run `salt-master` as root.
2. Run `salt-master` and `salt-syndic` as "salt", fix up permissions on the various dirs normally used by the minion so that the "salt" user can read them (we should document in detail). In this scenario you can't run a `salt-minion` normally on the master (you can if you setup an entirely separate config/root_dir for it).

In terms of an actual fix, I think I'd like to take a look at separating out the config and filesystem usage of `salt-syndic` so it's independent of the `salt-minion` on the master. As I said, I don't have much experience with syndic, but looking briefly at the code, I don't think there's any reason it should need the same privs as a normal minion.

**Comment by dmurphy18** [comment]
@hurzhurz This issues is being addressed by PR https://github.com/saltstack/salt/pull/66218 which affects permissions, user, etc.

**Comment by hurzhurz** [comment]
@dmurphy18 While I'm sure #66218 will help me in other regards, it doesn't look like it really fixes this issue.

I just took the packages from the current nighty build to test it again: [salt-3006.8+310.g962f7529fb-nightly-repo-deb-ubuntu-22.04-x86_64.zip](https://github.com/saltstack/salt/actions/runs/10068288198/artifacts/1732973799)
I guess they should already include the changes from #66218.

Result is that master and syndic still fight about the ownership of /var/cache/salt/master/proc.
And as before, with the default permissions of 755, the master (running as "salt") can't write there anymore after the syndic changes the owner to "root".

Though at the moment I can live with manually changing the permissions as a workaround (`chmod 775 /var/cache/salt/master/proc`).

**Comment by dmurphy18** [comment]
Will take another look, thought I had master and syndic with the same ownership and permissions.

**Comment by dmurphy18** [comment]
@hurzhurz Tested recent built with fixes #66218 included and it looks good after install salt-master, and then salt-syndic on a clean system - Rocky Linux 9 (can get past firewalls on Rocky 9, but not Ubuntu 22.04, have to test that outside of office)
```
-rw-r--r--.  1 david david  47M Jul 24 13:33 salt-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 223K Jul 24 13:33 salt-api-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 227K Jul 24 13:33 salt-cloud-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 1.9M Jul 24 13:33 salt-debuginfo-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 3.5M Jul 24 13:33 salt-master-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 254K Jul 24 13:33 salt-minion-3006.8+310.ga9b33feebf-0.x86_64.rpm
-rw-r--r--.  1 david david 225K Jul 24 13:33 salt-ssh-3006.8+310.ga9b33feebf-0.x86_64.rpm
drwxr-xr-x.  2 david david 4.0K Jul 24 13:33 .
-rw-r--r--.  1 david david 223K Jul 24 13:33 salt-syndic-3006.8+310.ga9b33feebf-0.x86_64.rpm
drwx------. 15 david david 4.0K Jul 24 13:35 ..
[root@dhcp-10-47-15-216 test]# l /var/cache/salt/
total 4.0K
drwxr-xr-x. 18 root root 4.0K Jul 24 13:35 ..
drwxr-xr-x.  3 root root   20 Jul 24 13:35 .
drwxr-x---.  8 salt salt   86 Jul 24 13:35 master
[root@dhcp-10-47-15-216 test]# l /var/cache/salt/master/
total 0
drwxr-x---. 2 salt salt  6 Jul 23 10:09 tokens
drwxr-x---. 2 salt salt  6 Jul 23 10:09 syndics
drwxr-x---. 2 salt salt  6 Jul 23 10:09 roots
drwxr-x---. 2 salt salt  6 Jul 23 10:09 queues
drwxr-x---. 2 salt salt  6 Jul 23 10:09 proc
drwxr-x---. 2 salt salt  6 Jul 23 10:09 jobs
drwxr-xr-x. 3 root root 20 Jul 24 13:35 ..
drwxr-x---. 8 salt salt 86 Jul 24 13:35 .
[root@dhcp-10-47-15-216 test]# l /var/cache/salt/master/
total 0
drwxr-x---. 2 salt salt  6 Jul 23 10:09 tokens
drwxr-x---. 2 salt salt  6 Jul 23 10:09 syndics
drwxr-x---. 2 salt salt  6 Jul 23 10:09 roots
drwxr-x---. 2 salt salt  6 Jul 23 10:09 queues
drwxr-x---. 2 salt salt  6 Jul 23 10:09 proc
drwxr-x---. 2 salt salt  6 Jul 23 10:09 jobs
drwxr-xr-x. 3 root root 20 Jul 24 13:35 ..
drwxr-x---. 8 salt salt 86 Jul 24 13:35 .
[root@dhcp-10-47-15-216 test]# 
```
second listing is after installing syndic.

**Comment by dmurphy18** [comment]
On Ubuntu 24.04 arm64
```
root@ip-172-31-51-6:/home/ubuntu/test# l /var/cache/salt/master/
total 40K
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 tokens
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 syndics
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 queues
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 proc
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 jobs
drwxr-xr-x 4 root root 4.0K Jul 24 22:30 ..
-rw------- 1 salt root   76 Jul 24 22:30 .salt_key
-rw------- 1 root root   76 Jul 24 22:30 .root_key
drwxr-xr-x 8 salt salt 4.0K Jul 24 22:30 .
drwxr-xr-x 2 salt salt 4.0K Jul 24 22:30 roots
root@ip-172-31-51-6:/home/ubuntu/test# # salt-master installed
root@ip-172-31-51-6:/home/ubuntu/test# 
root@ip-172-31-51-6:/home/ubuntu/test# systemctl status salt-syndic
● salt-syndic.service - The Salt Master Server
     Loaded: loaded (/usr/lib/systemd/system/salt-syndic.service; enabled; preset: enabled)
     Active: active (running) since Wed 2024-07-24 22:33:10 UTC; 32s ago
       Docs: man:salt-syndic(1)
             file:///usr/share/doc/salt/html/contents.html
             https://docs.saltproject.io/en/latest/contents.html
   Main PID: 3182 (/opt/saltstack/)
      Tasks: 1 (limit: 482)
     Memory: 42.9M (peak: 89.3M)
        CPU: 994ms
     CGroup: /system.slice/salt-syndic.service
             └─3182 "/opt/saltstack/salt/bin/python3.10 /usr/bin/salt-syndic MainProcess"

Jul 24 22:33:09 ip-172-31-51-6 systemd[1]: Starting salt-syndic.service - The Salt Master Server...
Jul 24 22:33:10 ip-172-31-51-6 systemd[1]: Started salt-syndic.service - The Salt Master Server.
Jul 24 22:33:12 ip-172-31-51-6 salt-syndic[3182]: [ERROR   ] DNS lookup or connection check of 'masterofmasters' failed.
Jul 24 22:33:12 ip-172-31-51-6 salt-syndic[3182]: [ERROR   ] Master hostname: 'masterofmasters' not found or not responsive. Retrying in 30 seconds
Jul 24 22:33:42 ip-172-31-51-6 salt-syndic[3182]: [ERROR   ] DNS lookup or connection check of 'masterofmasters' failed.
Jul 24 22:33:42 ip-172-31-51-6 salt-syndic[3182]: [ERROR   ] Master hostname: 'masterofmasters' not found or not responsive. Retrying in 30 seconds
root@ip-172-31-51-6:/home/ubuntu/test# l /var/cache/salt/master/
total 40K
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 tokens
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 syndics
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 queues
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 proc
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 jobs
drwxr-xr-x 4 root root 4.0K Jul 24 22:30 ..
-rw------- 1 salt root   76 Jul 24 22:30 .salt_key
-rw------- 1 root root   76 Jul 24 22:30 .root_key
drwxr-xr-x 8 salt salt 4.0K Jul 24 22:30 .
drwxr-xr-x 2 salt salt 4.0K Jul 24 22:30 roots
root@ip-172-31-51-6:/home/ubuntu/test# 
```

So after installing salt-syndic, /var/cache/salt/master/proc appears to have the correct ownership etc.

Perhaps you mis-installed. Salt 2006.9 is coming soon, dealing with a Windows 2022 issue,once that is resolved, it should be soon after, and you can retry the issue then.
The tests above where on a clean AWS VM Ubuntu 24.04 arm64 platform.

**Comment by hurzhurz** [comment]
@dmurphy18 According to the logs, your syndic wasn't able to connect to the masterofmasters. I guess you haven't setup one for this test?
In my test setup, I actually made a complete little setup with two VMs/containers, one as a masterofmasters and the other as a syndic.

So now I tested what happens if I stop the master service on the masterofmasters. Then indeed the syndic doesn't change the dir ownership of /var/cache/salt/master/proc.
It looks like it does it in the moment when it successfully connects to the masterofmasters.

**Comment by dmurphy18** [comment]
@hurzhurz No I did not setup a master of masters. So this looks like an operational issue, rather than a packaging / install issue, since you report it happening when the syndic connects to the master of masters.

Thought this was a packaging / install issue, since the permissions/groups problems with salt user and group introduced in 3006.0 has had a number of problems, that were contributing to the above problems. Those now fixed, I will look at  the operational aspect of this, as to what, if anything, is changing the user/group dynamically, which it probably should not be doing, wondering if blindly creating the file, running as root, when it checking if file already exists and using it (speculation).

**Comment by dmurphy18** [comment]
Setup masterofmaster with syndic on two VMs (Ubuntu 24.04 arm64) and did not see any issue with ownership
Master Of Master
```
root@ip-172-31-53-66:/home/ubuntu/test# l /var/cache/salt/master/
total 40K
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 tokens
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 syndics
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 queues
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 proc
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 jobs
drwxr-xr-x 2 salt salt 4.0K Jul 25 19:42 roots
drwxr-xr-x 4 root root 4.0K Jul 25 19:53 ..
-rw------- 1 root root   76 Jul 25 20:02 .root_key
-rw------- 1 salt root   76 Jul 25 20:02 .salt_key
drwxr-xr-x 8 salt salt 4.0K Jul 25 20:02 .
root@ip-172-31-53-66:/home/ubuntu/test# l /var/run/salt*
-rw-r--r-- 1 salt salt  4 Jul 25 20:02 /var/run/salt-master.pid

/var/run/salt:
total 0
drwxr-xr-x  3 root root  60 Jul 25 19:47 .
drwxr-xr-x 29 root root 940 Jul 25 19:53 ..
drwxr-xr-x  2 salt salt 120 Jul 25 20:02 master
root@ip-172-31-53-66:/home/ubuntu/test# ps -ef | grep salt
salt        3763       1  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master MainProcess
salt        3831    3763  0 20:02 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master PubServerChannel._publish_daemon
salt        3832    3763  0 20:02 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master EventPublisher
salt        3835    3763  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master Maintenance
salt        3836    3763  0 20:02 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer ReqServer_ProcessManager
salt        3837    3763  0 20:02 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master FileServerUpdate
salt        3838    3836  0 20:02 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorkerQueue
salt        3839    3836  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-0
salt        3841    3836  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-1
salt        3842    3836  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-2
salt        3843    3836  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-3
salt        3850    3836  0 20:02 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-4
root        4549    1699  0 20:07 pts/1    00:00:00 grep --color=auto salt
root@ip-172-31-53-66:/home/ubuntu/test# 
```

On the syndic, after connection with MoM
```
Jul 25 20:08:05 ip-172-31-50-187 systemd[1]: Starting salt-master.service - The Salt Master Server...
Jul 25 20:08:06 ip-172-31-50-187 systemd[1]: Started salt-master.service - The Salt Master Server.
root@ip-172-31-50-187:/home/ubuntu# ps -ef | grep salt
salt        4234       1 26 20:08 ?        00:00:02 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master MainProcess
root        4238       1 13 20:08 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-syndic MainProcess
root        5036    1558  0 20:08 pts/1    00:00:00 grep --color=auto salt
root@ip-172-31-50-187:/home/ubuntu# l /var/cache/salt/master/
total 40K
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 tokens
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 syndics
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 queues
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 proc
drwxr-xr-x 2 salt salt 4.0K Jul 23 15:48 jobs
drwxr-xr-x 4 root root 4.0K Jul 25 19:43 ..
drwxr-xr-x 2 salt salt 4.0K Jul 25 19:44 roots
-rw------- 1 salt root   76 Jul 25 20:08 .salt_key
-rw------- 1 root root   76 Jul 25 20:08 .root_key
drwxr-xr-x 8 salt salt 4.0K Jul 25 20:08 .
root@ip-172-31-50-187:/home/ubuntu# l /var/run/salt*
-rw-r--r-- 1 root root  4 Jul 25 20:08 /var/run/salt-syndic.pid
-rw-r--r-- 1 salt salt  4 Jul 25 20:08 /var/run/salt-master.pid

/var/run/salt:
total 0
drwxr-xr-x  2 root root  40 Jul 25 19:47 minion
drwxr-xr-x  4 root root  80 Jul 25 19:47 .
drwxr-xr-x 28 root root 920 Jul 25 20:08 ..
drwxr-xr-x  2 salt salt 120 Jul 25 20:08 master
root@ip-172-31-50-187:/home/ubuntu# 
```

Did notice on the syndic
```
-rw-r--r-- 1 root root  4 Jul 25 20:08 /var/run/salt-syndic.pid
```
and wonder if that should be salt:salt, but that is a different matter.

For now with 3006.9 coming soon (one last test to pass - Windows download test), I think you should wait on this and retry when 3006.9 is released, to see if the issue still occurs.

**Comment by hurzhurz** [comment]
It is strange that you don't see that issue.
Maybe another key element of my test setup is the scheduler, mention in my initial comment. This produces a bit of activity.
Anyway, I will test it again with 3006.9 and let you know.

And for the PID file... I guess root:root is to be expected as the syndic is actually running as root?
Of course it would be great if it could run as user salt too, like the master does, but as @barneysowood pointed out, this is probably complicated.

**Comment by hurzhurz** [comment]
ok, I have now tested it again with 3006.9 on Ubuntu 24.04 (amd64). Same as before.

So, this is exactly how I created my test setup:
```
incus launch images:ubuntu/24.04 test-mom
incus exec test-mom bash
---
echo "deb [trusted=yes] https://repo.saltproject.io/salt/py3/ubuntu/24.04/amd64/minor/3006.9 noble main" > /etc/apt/sources.list.d/salt.list
apt-get update && apt-get -y install salt-master salt-minion
echo "master: localhost" >> /etc/salt/minion
echo "order_masters: True" >> /etc/salt/master
```
```
incus launch images:ubuntu/24.04 test-syndic
incus exec test-syndic bash
---
echo "deb [trusted=yes] https://repo.saltproject.io/salt/py3/ubuntu/24.04/amd64/minor/3006.9 noble main" > /etc/apt/sources.list.d/salt.list
apt-get update && apt-get -y install salt-master salt-syndic salt-minion
echo "master: localhost" >> /etc/salt/minion
echo "syndic_master: 10.179.97.123" >> /etc/salt/master
```
After that, I of course also restarted all salt services, accepted all keys and verified the basic function with a `salt \* test.ping` on both.

And here is the result from checking the ownership:

```
### ownership already "root:salt":
root@test-syndic:~# ls -lah /var/cache/salt/master/proc
total 8.0K
drwxr-xr-x 2 root salt 4.0K Jul 29 07:42 .
drwxr-xr-x 9 salt salt 4.0K Aug  1 07:38 ..

### stopping master&syndic:
root@test-syndic:~# systemctl stop salt-syndic
root@test-syndic:~# systemctl stop salt-master

### first starting master:
root@test-syndic:~# systemctl start salt-master
### ownership is now "salt:salt":
root@test-syndic:~# ls -lah /var/cache/salt/master/proc
total 8.0K
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 .
drwxr-xr-x 9 salt salt 4.0K Aug  1 07:40 ..

### starting syndic:
root@test-syndic:~# systemctl start salt-syndic
### after some seconds (probably after syndic has connected), ownership is "root:salt" again:
root@test-syndic:~# ls -lah /var/cache/salt/master/proc
total 8.0K
drwxr-xr-x 2 root salt 4.0K Jul 29 07:42 .
drwxr-xr-x 9 salt salt 4.0K Aug  1 07:40 ..
```

**Comment by dmurphy18** [comment]
@hurzhurz Something in your environment is messing with things
```
### ownership already "root:salt":
root@test-syndic:~# ls -lah /var/cache/salt/master/proc
total 8.0K
drwxr-xr-x 2 root salt 4.0K Jul 29 07:42 .
drwxr-xr-x 9 salt salt 4.0K Aug  1 07:38 ..
```
Since when I installed salt-master and salt-syndic on seperate clean VM nodes for Ubuntu, for ```/var/cache/salt/master/proc``` I had ```salt:salt```.

I got clean VMs , used the install instructions from https://docs.saltproject.io/salt/install-guide/en/3006/topics/install-by-operating-system/ubuntu.html#install-salt-on-ubuntu-24-04-noble-amd64 and used the appropriate ``` apt install salt-master``` and ```apt install salt-syndic``` 

Note: did spot mistake in install instructions for Ubuntu 24.04, use of jammy instead of noble, being fixed.

This from a clean AWS Ubuntu 24.04 node (amd64)
```

root@ip-172-31-7-153:/home/ubuntu# ps -ef | grep salt
salt        2571       1  2 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master MainProcess
salt        2737    2571  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master PubServerChannel._publish_daemon
salt        2738    2571  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master EventPublisher
salt        2747    2571  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master Maintenance
root        2749       1  2 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-syndic MainProcess
salt        2750    2571  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer ReqServer_ProcessManager
salt        2753    2750  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorkerQueue
salt        2754    2750  3 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-0
salt        2755    2750  3 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-1
salt        2756    2750  3 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-2
salt        2757    2571  0 19:20 ?        00:00:00 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master FileServerUpdate
salt        2758    2750  3 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-3
salt        2759    2750  3 19:20 ?        00:00:01 /opt/saltstack/salt/bin/python3.10 /usr/bin/salt-master ReqServer MWorker-4
root        3544    1195  0 19:20 pts/1    00:00:00 grep --color=auto salt
root@ip-172-31-7-153:/home/ubuntu# ls -alrth --color=auto /var/cache/salt/master
total 40K
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 tokens
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 syndics
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 queues
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 proc
drwxr-xr-x 2 salt salt 4.0K Jul 29 07:42 jobs
drwxr-xr-x 4 root root 4.0K Aug  1 19:20 ..
-rw------- 1 salt root   76 Aug  1 19:20 .salt_key
-rw------- 1 root root   76 Aug  1 19:20 .root_key
drwxr-xr-x 8 salt salt 4.0K Aug  1 19:20 .
drwxr-xr-x 2 salt salt 4.0K Aug  1 19:20 roots
root@ip-172-31-7-153:/home/ubuntu# 
```
Didn't even touch configuration files etc.

I would check how you install in your environment, use of ```incus``` and how it affects permissions

**Comment by dmurphy18** [comment]
Try using an AWS Ubuntu 24.04 VM to check for yourself, rather than relying on incus

**Comment by dmurphy18** [comment]
@hurzhurz You are forgetting that I already did that, setting up a Mom, syndic, etc. on AWS VMs and did not see the problem, see https://github.com/saltstack/salt/issues/65095#issuecomment-2251314834

So this is going around in circles, is there some other information which can throw light on how to make the issue occur, given all steps to date have failed trying to reproduce the issue

**Comment by hurzhurz** [comment]
Yeah, I remember that you tried it. And it is very strange and a bit frustrating that you didn't see the problem while I have no trouble at all to reproduce it.
Can you share what exact configuration you did thereby? Did it differ from my config in the setup steps in https://github.com/saltstack/salt/issues/65095#issuecomment-2262297284)?

I will also try to do some more testing tomorrow.
Maybe also try to narrow it down when exactly the syndic process changes the ownership in my setup: I think about setting the log level to debug or trace and try find the position where it roughly happens.

**Comment by hurzhurz** [comment]
Ok, so I did a little trick to find the code where the owner change happens:
Before starting the syndic service, I made /var/cache/salt/master a read-only mount point, so it won't be able to change anything to provoke an error:
```
mv /var/cache/salt/master /var/cache/salt/master-org
mkdir /var/cache/salt/master
mount -o bind,ro /var/cache/salt/master-org /var/cache/salt/master`
```

And that worked perfectly, it causes this exception:

```
2024-08-02 07:47:34,475 [salt.minion      :3550][CRITICAL][59036] Unexpected error while connecting to 10.179.97.123
Traceback (most recent call last):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 3526, in _connect_syndic
    yield syndic.connect_master(failed=failed)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/gen.py", line 1056, in run
    value = future.result()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/gen.py", line 1064, in run
    yielded = self.gen.throw(*exc_info)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1404, in connect_master
    yield self._post_master_init(master)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/gen.py", line 1056, in run
    value = future.result()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/concurrent.py", line 249, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/ext/tornado/gen.py", line 1070, in run
    yielded = self.gen.send(value)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1443, in _post_master_init
    self._setup_core()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 3002, in _setup_core
    self.proc_dir = get_proc_dir(self.opts["cachedir"], uid=uid)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 313, in get_proc_dir
    os.chown(fn_, uid, gid)
OSError: [Errno 30] Read-only file system: '/var/cache/salt/master/proc'
```
So as suspected, it happens exactly here when connecting (def _connect_syndic): https://github.com/saltstack/salt/blob/v3006.9/salt/minion.py#L313

**Comment by dmurphy18** [comment]
@hurzhurz Thanks for identifying the line, will take a look at this next week, given Friday afternoon, once finished with current task

**Comment by dmurphy18** [comment]
Closing since PR https://github.com/saltstack/salt/pull/66803 is merged into 3006.x branch
